### PR TITLE
add v2 recipe editing capabilities for bumping version & build number

### PR DESCRIPTION
**Pull request by wolfv**
Hi @beckermr - this is the initial PR to port over the functionality for the recipe editing for v2 recipes. 

I'll add some tests now.

### CODE REVIEW

**Comment by wolfv** [comment]
@beckermr - I ported all recipe editing here. It's a bit involved because the tool does the following:

- evaluate the variables in the context using Jinja
- find all the sources, irregardless of `if/else` branches
- render the version _with_ the variables loaded from the context
- edit the recipe to insert the new hashes & version

lmk what you think! :)

**Comment by beckermr** [comment]
Yep. If you look at the existing code it detects a bunch of errors and outputs an error message etc. we need to match that api.

**Comment by beckermr** [comment]
Also the hashing bits should use the util in the repo. We've had issues with hashing stalling for some URLs and killing the bot. I'll leave more comments later. Sorry for the terse comments.

**Comment by wolfv** [comment]
@beckermr - I just tackled the API of `update_version`, will make sure that the build number update also matches. Reused the existing hashing function.

File: conda_forge_tick/migrators/version.py [hunk 1]
```diff
@@ -195,23 +195,31 @@ def migrate(
...
-                with open("meta.yaml", "w") as fp:
-                    fp.write(updated_meta_yaml)
-                self.set_build_number("meta.yaml")
+            if len(errors) == 0 and updated_meta_yaml is not None:
+                meta_yaml.write_text(updated_meta_yaml)
+                self.set_build_number(meta_yaml)
 
-            return super().migrate(recipe_dir, attrs)
-        else:
+                return super().migrate(recipe_dir, attrs)
```
Comment (beckermr) [inline conda_forge_tick/migrators/version.py:214]: This return clause needs be added for v2 or combined into a single set of statements at the end of the method.

File: conda_forge_tick/update_recipe/v2/build_number.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    Returns:
+    --------
+    The updated recipe as a string.
+    """
+    data = _load_yaml(file)
+    build_number_modified = _update_build_number_in_context(data, new_build_number)
+    if not build_number_modified:
+        _update_build_number_in_recipe(data, new_build_number)
+
+    return _dump_yaml_to_str(data)
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/build_number.py:61]: People can and have mixed setting build numbers in both the jinja2 vars / context and explicitly in the recipe. I know it is awful. We should update the code to do both and check if the build number in the non-context part is templated or not before updating.

File: conda_forge_tick/update_recipe/v2/context.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import jinja2
+
+
+def load_recipe_context(
+    context: dict[str, str], jinja_env: jinja2.Environment
+) -> dict[str, str]:
+    """
+    Load all string values from the context dictionary as Jinja2 templates.
+    Use linux-64 as default target_platform, build_platform, and mpi.
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/context.py:9]: I'm concerned that assuming linux as the default will result in invalid updates for some recipes.

File: conda_forge_tick/update_recipe/v2/jinja/filters.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    major = parts[0] if len(parts) > 0 else ""
+    minor = parts[1] if len(parts) > 1 else ""
+    return f"{major}{minor}"
+
+
+def _bool(value: str) -> bool:
+    return bool(value)
+
+
+def _split(s: str, sep: str = " ") -> list[str]:
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/jinja/filters.py:26]: Setting this default to `None` would match standard python behavior. IDK if we should do that or not, but it seems like it might be a good idea.

File: conda_forge_tick/update_recipe/v2/jinja/jinja.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    Target platform, build platform, and mpi are set to linux-64 by default.
+    """
+    env = SandboxedEnvironment(
+        variable_start_string="${{",
+        variable_end_string="}}",
+        trim_blocks=True,
+        lstrip_blocks=True,
+        autoescape=jinja2.select_autoescape(default_for_string=False),
+        undefined=_MissingUndefined,
+    )
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/jinja/jinja.py:43]: Can you add a comment on these options, why they are set, and how that relates to the supported features for the v2 recipe?

Also, will this parser fail on `if` conditions, `for` loops, or other non-supported jinja2 filters / syntax?

File: conda_forge_tick/update_recipe/v2/jinja/jinja.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    return env
+
+
+def load_recipe_context(
+    context: dict[str, str], jinja_env: jinja2.Environment
+) -> dict[str, str]:
+    """
+    Load all string values from the context dictionary as Jinja2 templates.
+    Use linux-64 as default target_platform, build_platform, and mpi.
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/jinja/jinja.py:84]: Again I am concerned about the defaults here for platform etc.

File: conda_forge_tick/update_recipe/v2/source.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        return
+
+    outputs = visit_conditional_list(outputs, None)
+    for output in outputs:
+        sources = output.get("source", None)
+        sources = typing.cast(ConditionalList[Source], sources)
+        if sources is None:
+            continue
+        source_list = visit_conditional_list(sources, None)
+        yield from source_list
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/source.py:50]: Somehow it escaped me that outputs can have their own sources now. That will be fun.

File: conda_forge_tick/update_recipe/v2/version.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        return f"{self.hash_type}: {self.hash_value}"
+
+
+def _has_jinja_version(url: str) -> bool:
+    """Check if the URL has a jinja `${{ version }}` in it."""
+    pattern = r"\${{\s*version"
+    return re.search(pattern, url) is not None
+
+
+def update_hash(source: Source, url: str, hash_: Hash | None) -> None:
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/version.py:53]: Some people will set the hash in the context section. yay edge cases!

File: conda_forge_tick/update_recipe/v2/version.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+        if not _has_jinja_version(url):
+            continue
+
+        template = env.from_string(url)
+        rendered_url = template.render(context_variables)
+        try:
+            update_hash(source, rendered_url, hash_)
+        except HashError:
+            return None, set(f"Could not retrieve hash for '{rendered_url}'")
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/version.py:133]: Instead of an early return, the bot should collect all of the errors and return them once. These error strings appear on the status page.

File: conda_forge_tick/update_recipe/v2/version.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    def __init__(self, message: str = "Could not update version") -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class HashError(Exception):
+    def __init__(self, url: str) -> None:
+        self.url = url
+        self.message = f"Could not hash {url}"
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/version.py:34]: It is useful to add the templated string that was rendered and what version in that string was tried to this error so that folks can diagnose issues.

File: conda_forge_tick/update_recipe/v2/version.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    for source in get_all_sources(data):
+        # render the whole URL and find the hash
+        if "url" not in source:
+            continue
+
+        url = source["url"]
+        if isinstance(url, list):
+            url = url[0]
+
+        if not _has_jinja_version(url):
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/version.py:125]: URLs can have other templated items that change (like version subparts) and so will need to updated.

File: conda_forge_tick/update_recipe/v2/yaml.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import io
+from pathlib import Path
+
+from ruamel.yaml import YAML
+
+yaml = YAML()
+yaml.preserve_quotes = True
+yaml.width = 4096
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/yaml.py:8]: This is pretty big. The bot uses 320 elsewhere - https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/recipe_parser/_parser.py#L46

File: conda_forge_tick/update_recipe/v2/version.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            continue
+
+        template = env.from_string(url)
+        rendered_url = template.render(context_variables)
```
Comment (beckermr) [inline conda_forge_tick/update_recipe/v2/version.py:129]: The bot tries a series of url template modifications in order to get version updates to work. People do things like add / remove `v`, there are multiple pypi urls etc. We need to use those here.

**Review by beckermr — CHANGES_REQUESTED** [review]
I left a bunch of comments. One more is that the tests are not complete in that they don't touch a lot of the edge cases. 

This PR is getting pretty big, so my suggestion is that we split to three or four PRs: 1) add the parser + tests for it, 2) the build number bump, 3) the version change + basic tests, and 4) version tests for edge cases.

File: conda_forge_tick/update_recipe/v2/context.py (hunk 3 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/context.py:9]: I understand where you are coming from but we do ignore the selectors in the version updating code (e.g. we go down both branches, `then` and `else`.

However, we do use the value for templating, so people could theoretically do something like 

```yaml
source:
  url: https://foo.${{ "linux" if linux else "bla" }}/...
  sha256: ${{ "foo" ...
```

But that doesn't seem very likely and also buggy so I don't think we'll see that a lot in the wild.

File: conda_forge_tick/update_recipe/v2/jinja/filters.py (hunk 4 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/jinja/filters.py:26]: We should match the `minijinja` behavior which is to split on any whitespace (and also the same as the default in Python I think). So that's a good point! Thanks!

File: conda_forge_tick/update_recipe/v2/source.py (hunk 7 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/source.py:50]: Wasn't this always the case? I think it was always the case.

File: conda_forge_tick/update_recipe/v2/version.py (hunk 8 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:53]: Yeah it's pretty discouraged by the schema right now and I don't see the benefit for the users to set the hash in the context. We just inject it in the source and the old hash will persist in the context for now. 
If someone gets annoyed we can modify this.

File: conda_forge_tick/update_recipe/v2/jinja/jinja.py (hunk 5 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/jinja/jinja.py:43]: Is there a way we can make it fail and output an error?

File: conda_forge_tick/update_recipe/v2/yaml.py (hunk 12 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/yaml.py:8]: I was under the impression that we would never want word wrapping where it wasn't done by the user, but sure, can use the suggestion :)

File: conda_forge_tick/update_recipe/v2/version.py (hunk 13 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:129]: Yeah, that would all be done in the `context` or in the rendering process. There is no functional difference here.

**Comment by wolfv** [comment]
Thanks for the comments @beckermr - I'll try to get to it asap.

File: conda_forge_tick/update_recipe/v2/version.py (hunk 9 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/version.py:133]: Many recipes have more than once source.

File: conda_forge_tick/update_recipe/v2/version.py (hunk 8 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/version.py:53]: This could break some recipes and people will be even more annoyed by that. We need to fix it.

File: conda_forge_tick/update_recipe/v2/yaml.py (hunk 12 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/yaml.py:8]: I tuned this carefully after several complaints to do enough wrapping to make recipes legible, but not enough to annoy folks. :)

File: conda_forge_tick/update_recipe/v2/version.py (hunk 9 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:133]: true! let me double check that case.

File: conda_forge_tick/update_recipe/v2/version.py (hunk 13 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:129]: Wait, you mean "people" as in upstream maintainers doing that?

Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/version.py:129]: I mean upstream changes the urls and the bot tries to account for it. See here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/url_transforms.py

**Comment by beckermr** [comment]
No rush!

File: conda_forge_tick/update_recipe/v2/version.py (hunk 13 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py]: Which also lets me wonder if we need to adjust the code somewhere else where the different versions are being tried out :)

File: conda_forge_tick/update_recipe/v2/jinja/filters.py (hunk 4 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/jinja/filters.py:26]: Done!

File: conda_forge_tick/update_recipe/v2/context.py (hunk 3 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/context.py:9]: I believe that doesn't work in the context section. However you _can_ use templating and inline Jinja expressions, such as:

```
context:
  version: "1.2.3"
  version_string: "v${{ version if linux else "foo" }}"
```

File: conda_forge_tick/update_recipe/v2/version.py (hunk 13 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/version.py]: The version ticker thing does this too, but it does not record which url variant worked. Hence it is done here again. It is possible that the recipe can change between when the version is found and when we update the recipe as well.

File: conda_forge_tick/update_recipe/v2/context.py (hunk 3 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/context.py:9]: right so we need to make this correct.

File: conda_forge_tick/migrators/version.py (hunk 1 above)
Comment (wolfv) [reply conda_forge_tick/migrators/version.py:214]: thanks, fixed.

File: conda_forge_tick/update_recipe/v2/context.py (hunk 3 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/context.py:9]: I think it's bad practice in many ways. 

Can I instead add a lint on the sha256 hash being in the proper spot (NOT in context)?
Then people would not / could not do many shenanigans with the URL either.

Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/context.py:9]: That won't change other items that might be platform dependent in the context section.

Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/context.py:9]: But the SHA hash will make it impossible, thankfully :)

Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/context.py:9]: Nope. Folks could define parts of urls in the context section that are platform dependent that are not hashes. So when the bot comes through to render those urls and parts, it will compute the wrong hash since it might have a valid, but incorrect url for that platform.

Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/context.py:9]: @beckermr - i sorta get where you come from but is it really necessary to care about all of these edge cases on day one?

It is also not a very clever way to write the recipe. You would have to "double-if" the source.

E.g.

```yaml
context:
  foo: ${{ "bla" if linux else "foo" }}

source:
  - if: linux
    then:
      url: ${{ foo }}
      sha256: 1abcd
    else:
      url: ${{ foo }}
      sha256: 2bcde
```

So I don't think many people would wanna write their sources like this (doesn't make much logical sense).

File: conda_forge_tick/update_recipe/v2/version.py (hunk 8 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:46]: I made a lint to prevent this in conda-smithy: https://github.com/conda-forge/conda-smithy/pull/2028

Still needs tests etc. but I think that's the better way forward.

File: conda_forge_tick/update_recipe/v2/context.py (hunk 3 above)
Comment (beckermr) [reply conda_forge_tick/update_recipe/v2/context.py:9]: Yes we should care. My suggested procedure for this work was to abstract the current version update algorithm to be independent of the recipe format backend and then use that. It should handle most, if not all, of the cases I am bringing up.

File: conda_forge_tick/update_recipe/v2/version.py (hunk 8 above)
Comment (wolfv) [reply conda_forge_tick/update_recipe/v2/version.py:46]: @beckermr how would it break a recipe though. You would have:

```
context:
  sha256: foo

source:
  url: bla
  sha256: "1234..."
```

At some point someone would come and clean up the left-over hash from the context.

It would only break if the SHA hash was part of the URL. But in that case, the autotick updater has basically no chance anyways :)

**Comment by wolfv** [comment]
I wanted to quickly type out what a v2 recipe looks like:

```yaml
context:
  version: "0.9.6"

package:
  name: bottom
  version: ${{ version }}
  
source:
  url: https://github.com/ClementTsang/bottom/archive/refs/tags/${{ version }}.tar.gz
  sha256: 202130e0d7c362d0d0cf211f6a13e31be3a02f13f998f88571e59a7735d60667

build:
  script:
    - if: unix
      then:
        - cargo install --locked --root $PREFIX --path . --no-track
      else:
        - cargo install --locked --root %PREFIX% --path . --no-track

requirements:
  build:
    - ${{ compiler("rust") }}

tests:
  - script:
      - btm --help 
```

Now if you had multiple sources, or conditional sources, you are still forced to spell them out. E.g.

```yaml
source:
  - if: linux
    then:
      url: https://github.com/ClementTsang/bottom/archive/refs/tags/${{ version }}-linux.tar.gz
      sha256: 202130e0d7c362d0d0cf211f6a13e31be3a02f13f998f88571e59a7735d60667
  - if: win
    then:
      url: https://github.com/ClementTsang/bottom/archive/refs/tags/${{ version }}-win.tar.gz
      sha256: 8a86c4eecf12446ff273afc03e1b3a09a911d0b7981db1af58cb45c439161295
  - if: osx
    then:
      url: https://github.com/ClementTsang/bottom/archive/refs/tags/${{ version }}-osx.tar.gz
      sha256: 688787d8ff144c502c7f5cffaafe2cc588d86079f9de88304c26b0cb99ce91c6
```

I looked at the code for the selector evaluation and it makes a lot of sense for meta.yamls'. 

But for the new recipe format it doesn't really make sense. The use case doesn't really exist because people wouldn't write it that way. You wouldn't even really use any platform specific jinja variables in the URLs etc. since you can just insert the proper URL (just include the version).

So IMO we can ignore this for the new recipes. I also do not (really not) see a good use case for templating the SHA hash.

Can someone come up with a good use case for templating that hash (for the new recipe format)? Maybe I am blind! :)

**Comment by wolfv** [comment]
Regarding the version variations - that makes sense. We only support a subset of what is being tried in the current munger but that should be fine. I'll try to get to it. There is also an item about the R-version needing some slight adjustment that we should handle.

**Comment by beckermr** [comment]
Yeah it's not really about what you think is a good or bad use of the format. If it is allowed, it will happen and the bot will have to deal with it. :/

**Comment by mgorny** [comment]
Hope you don't me asking but where are we standing here? Is there anything I could do to help here? I'm new to this codebase, but I think I can spend some time on it; just don't want to step on anyone's toes.

**Comment by wolfv** [comment]
Hi @mgorny - sorry for the slow response! 

This was working, but Isuru and Matt wanted to cover a number of special cases (e.g. templated SHA256 hashes).

I have the following prototype to render out the sources here: https://github.com/prefix-dev/rattler-build-conda-compat/pull/57

The idea is:

- we render against all variant configs and extract all URL sources
- check for updates and replace URLs and sha256 hashes as "text replacements" so that any comments stay intact, etc.

If you have some cycles to spend on this, that would be amazing. I added some tests in the PR linked that should show the URL source extraction to be working.

**Comment by mgorny** [comment]
Do I understand correctly that you're suggesting that I take the prototype code from https://github.com/prefix-dev/rattler-build-conda-compat/pull/57 and try to integrate it into this pull request? Or do you have something else in mind?

**Comment by ytausch** [comment]
Is this superseded by #3525?

**Comment by wolfv** [comment]
yep!
